src/database/sql/sql.go

```diff
@@ -909,17 +919,14 @@ func (db *DB) connectionCleaner(d time.Duration) {
 			return
 		}
 
-		expiredSince := nowFunc().Add(-d)
 		var closing []*driverConn
-		for i := 0; i < len(db.freeConn); i++ {
-			c := db.freeConn[i]
-			if c.createdAt.Before(expiredSince) {
+		freeConnLen := len(db.freeConnCh)
+		for i := 0; i < freeConnLen; i++ {
+			c := <- db.freeConnCh
+			if c.expired(d) {
 				closing = append(closing, c)
-				last := len(db.freeConn) - 1
-				db.freeConn[i] = db.freeConn[last]
-				db.freeConn[last] = nil
-				db.freeConn = db.freeConn[:last]
-				i--
+			} else {
+				db.freeConnCh <- c
```

Comment:
i worry about this change. i will think about this more. i'd feel comfortable if golang reviewers are okay with this.